wasm-node/src/node.rs

```diff
@@ -0,0 +1,187 @@
+use js_sys::Array;
+use libp2p::identity::Keypair;
+use libp2p::{identity, Multiaddr};
+use serde_wasm_bindgen::{from_value, to_value};
+use tracing::info;
+use wasm_bindgen::prelude::*;
+
+use crate::utils::Network;
+use crate::wrapper::libp2p::NetworkInfo;
+use crate::Result;
+
+#[wasm_bindgen(js_name = Node)]
+struct WasmNode {
+    node: Node<IndexedDbStore>,
+}
+
+#[wasm_bindgen(js_name = NodeConfig)]
+pub struct WasmNodeConfig {
+    pub network: Network,
+    #[wasm_bindgen(skip)]
+    pub genesis_hash: Option<Hash>,
+    #[wasm_bindgen(skip)]
+    pub p2p_local_keypair: Keypair,
+    #[wasm_bindgen(skip)]
+    pub p2p_bootnodes: Vec<Multiaddr>,
+}
+
+#[wasm_bindgen(js_class = Node)]
+impl WasmNode {
+    #[wasm_bindgen(constructor)]
+    pub async fn new(config: WasmNodeConfig) -> Self {
+        let network_id = network_id(config.network.into());
+        let store = IndexedDbStore::new(network_id).await.unwrap();
+        info!(
+            "Initialised store with head height: {:?}",
+            store.get_head_height()
+        );
+
+        let node = Node::new(NodeConfig {
+            network_id: network_id.to_string(),
+            genesis_hash: config.genesis_hash,
+            p2p_local_keypair: config.p2p_local_keypair,
+            p2p_bootnodes: config.p2p_bootnodes,
+            p2p_listen_on: vec![],
+            store,
+        })
+        .await
+        .context("Failed to start node")
+        .unwrap_throw();
+
+        Self { node }
+    }
+
+    pub fn local_peer_id(&self) -> String {
+        self.node.p2p().local_peer_id().to_string()
+    }
+
+    pub async fn wait_connected(&self) -> Result<()> {
+        Ok(self.node.p2p().wait_connected().await?)
+    }
+
+    pub async fn wait_connected_trusted(&self) -> Result<()> {
+        Ok(self.node.p2p().wait_connected_trusted().await?)
+    }
+
+    pub async fn network_info(&self) -> Result<NetworkInfo> {
+        Ok(self.node.p2p().network_info().await?.into())
+    }
+
+    pub async fn get_head_header(&self) -> Result<JsValue> {
+        let eh = self.node.p2p().get_head_header().await?;
+        Ok(to_value(&eh)?)
+    }
+
+    pub async fn get_header(&self, hash: &str) -> Result<JsValue> {
+        let hash: Hash = hash.parse()?;
+        let eh = self.node.p2p().get_header(hash).await?;
+        Ok(to_value(&eh)?)
+    }
+
+    pub async fn get_header_by_height(&self, height: u64) -> Result<JsValue> {
+        let eh = self.node.p2p().get_header_by_height(height).await?;
+        Ok(to_value(&eh)?)
+    }
+
+    pub async fn get_verified_headers_range(&self, from: JsValue, amount: u64) -> Result<Array> {
+        let header = from_value::<ExtendedHeader>(from)?;
+        let verified_headers = self
+            .node
+            .p2p()
+            .get_verified_headers_range(&header, amount)
+            .await?;
+
+        Ok(Array::from_iter(
+            verified_headers.iter().map(|v| to_value(v).unwrap_throw()),
+        ))
```

Comment:
I've wanted to avoid `collecting` here, because `Array` has constructor `from_iter`. Other ways I've tried to avoid `unwrap_throw`resulted in additional copy of the entire array getting created, as `Vec` cannot be returned directly.

Reply:
iirc `from_iter` comes from `FromIterator` trait and is what allows you to actually call `.collect()`. I see them equivalent except that `.collect()` is rather preferred unless some less common use cases

Reply:
https://doc.rust-lang.org/stable/std/iter/trait.FromIterator.html